### ISSUE TITLE
Combobox style corrupted when placed on a tab control in an accordian

### ISSUE THREAD

**Issue by PeterWoods16** [DevExpress/AjaxControlToolkit#282]
### The AJAX Control Toolkit \<16.1.1.0\>
Tested in 16.1.1.0 and 15.1.4.0
MS VS Com 2015 v14.0.25431.01 update 3
.Net Framework 4.6.01586
### The installation method
NuGet package manager console
     Resolving actions to install package 'AjaxControlToolkit.16.1.1'
     Resolved actions to install package 'AjaxControlToolkit.16.1.1'
     Retrieving package 'AjaxControlToolkit 16.1.1' from 'nuget.org'.
       GET https://api.nuget.org/packages/ajaxcontroltoolkit.16.1.1.nupkg
       OK https://api.nuget.org/packages/ajaxcontroltoolkit.16.1.1.nupkg 123ms
     Installing AjaxControlToolkit 16.1.1.
     Adding package 'AjaxControlToolkit.16.1.1' to folder 'C:\Work\Source\Web\Scratch\packages'
     Added package 'AjaxControlToolkit.16.1.1' to folder 'C:\Work\Source\Web\Scratch\packages'
     Added package 'AjaxControlToolkit.16.1.1' to 'packages.config'
     Successfully installed 'AjaxControlToolkit 16.1.1' to Scratch
     Executing nuget actions took 4.42 sec
     Time Elapsed: 00:00:05.3199981
     PM> 
### Minimal steps to reproduce the bug
This is the problem cut to the most basic HTML...
Basically, it is an Accordion control with a combo in the first pane and then a tab control that contains combo in the second pane.  A combo3 control is added to the second accordion pane outside of the scope of the tab control to illustrate that fact that without the tab everything is actually ok.

[Html.txt](https://github.com/DevExpress/AjaxControlToolkit/files/660606/Html.txt)

### Actual result
This is with IE11, but its the same with Chrome etc.
Everything is find as it loads up and shows the first accordion pane.

![image](https://cloud.githubusercontent.com/assets/24647830/21306449/f2032fc8-c5c8-11e6-98d8-34981299cdf7.png)

So then click on 'Header 2'

![image](https://cloud.githubusercontent.com/assets/24647830/21306495/2c212fde-c5c9-11e6-8dc7-e7af8a9f31eb.png)

So what this shows is the second combo in the tab control and a third combo outside the tab control.  The second combo has a corrupted style whereas the third combo is still ok.  Specifically, the corrupted combo2 has zero sized height and width for the button and list aspects of the control.

![image](https://cloud.githubusercontent.com/assets/24647830/21306667/1cfdf0a4-c5ca-11e6-8246-2e3851450b1f.png)

The above debugger output shows the width and height style values set to zero on combo2.

For comparison, here are the same styles set on combo3 correctly (outside of the tab control).

![image](https://cloud.githubusercontent.com/assets/24647830/21306802/b5145d2e-c5ca-11e6-8fc2-e921467a9339.png)

### Expected result
My expected result would be a combo styled exactly as combo1 and combo3 inside the tab control within the second accordion pane i.e.....

![image](https://cloud.githubusercontent.com/assets/24647830/21306890/19554258-c5cb-11e6-8971-e4991adbde3c.png)

### Browser(s) used
IE11 & Chrome 55.0.2883.87 m (64 bit)
### A site deployment method
I am just deploying to LOCALHOST at the moment .
Win 10, IIS Express.

**Comment by MikhailTymchukDX** [comment]
Hi @PeterWoods16 ,
Thank you for such a detailed report!

This issue occurs because of the following situation:
1. The height and width of the ComboBox button [is set only once during the control initialization](https://github.com/DevExpress/AjaxControlToolkit/blob/master/AjaxControlToolkit/Scripts/ComboBox.js#L234-L237).
1. `offsetHeight` and `offsetWidth` of a ComboBox's input field is always 0, because it is hidden inside a collapsed Accordion pane.

To work this around, please set the height and width of the button explicitly:
```js
<script type="text/javascript">
    Sys.Application.add_load(function() {
        var accordion = $find("<%= Accordion1.ClientID%>" + "_AccordionExtender");
        accordion.add_selectedIndexChanged(function(e) {
            if(e._selectedIndex === 1) {
                var combobox = $find("<%= ComboBox2.ClientID%>");
                combobox.get_buttonControl().style.height = combobox.get_textBoxControl().offsetHeight + 'px';
                combobox.get_buttonControl().style.width = combobox.get_buttonControl().style.height;
            }
        });
    });
</script>	
```

**Comment by PeterWoods16** [comment]
Thank you for your prompt response.
I should say from the outset - I usually confine myself to JS rather than JQuery so I had a bit of a learning curve to deploy your response.
First I had to insert the code at the end of the script to avoid a crash on the reference to Sys.Application.  Normally, I insert JS in the header?
Another issue were your <% ... %> references.  I think that's where I could have set up Accordion1.ClientID in the code behind to avoid the long references?  Was not quite sure about, that so I just hard coded the references in.

So here is what I ended up with...

[Html2.txt](https://github.com/DevExpress/AjaxControlToolkit/files/663077/Html2.txt)

The above code fixes the issue on the combo button, but of course does not address the issue with the optionlist.  The red ring illustrates the list dropped down!

![image](https://cloud.githubusercontent.com/assets/24647830/21343216/9c1e6788-c68d-11e6-8f31-1fca8968681c.png)

I started to extend your code toward fixing the option list but I was not too sure where the get_buttonControl() function had come from?  Is that an Ajax function call?  Clearly, I need an equivalent function to get the optionlist control so I can correct the height and width there too.
Once I have a handle on the control we are presumably going to need to set the optionlists height based on the number of list items?

Finally...
Do Codeplex regard this as a bug?
Is this a known issue?
Will it be fixed in Ajax 17?
Are any other controls affected in this way?  i.e. that render incorrectly when placed in a tab within an accordion.

Cheers once again,

Pete.

**Comment by PeterWoods16** [comment]
As a further update to the above, I have now located.

https://github.com/DevExpress/AjaxControlToolkit/wiki/ComboBox

And specifically...

optionListControl

An OptionList control.

Getter name: get_optionListControl()
Setter name: set_optionListControl(value)

And therefore have updated my code to...

[Html3.txt](https://github.com/DevExpress/AjaxControlToolkit/files/663207/Html3.txt)

I have set the height and width to 200px just to make sure I see a difference in the list size.  I can figure out how to get the proper values later!

But what happens now is (shown zoomed so we can see what's going on)...

Shown here is the FIRST click on the button (to open the list).

![image](https://cloud.githubusercontent.com/assets/24647830/21344532/1f3e9dd0-c694-11e6-8366-8f4a4a541410.png)

The SECOND click on the button is, as expected, "uneventful" as it just closes up the list (makes it invisible again).

Here is the THIRD click on the list (and how it shows for all subsequent button press 'list open clicks')...

![image](https://cloud.githubusercontent.com/assets/24647830/21344694/f2c8969c-c694-11e6-9806-4b5c5089547e.png)

The height does not alter, but strangely the width has now increased?

I suspect my code is in the wrong place here.  What must be happening, is that the listcontrol is being styled at the point of the button click event, so that it can be based on the contents (number of items) at the time of the click, rather than the form load (makes sense).

So if that is the case, how do we hook into that event to fix this?

**Comment by MikhailTymchukDX** [comment]
@PeterWoods16 
Thank you for providing the new info about an optionlist issue.

To fix both the button and optionlist layout, please use the following code:
```js
<script type="text/javascript">
	Sys.Application.add_load(function() {
		var accordion = $find("<%= Accordion1.ClientID%>" + "_AccordionExtender");
		accordion.add_selectedIndexChanged(function(e) {
			if(e._selectedIndex === 1) {
				var combobox = $find("<%= ComboBox2.ClientID%>");

				combobox.get_buttonControl().style.height = '';
				combobox.get_buttonControl().style.width = '';
				combobox.initializeButton();

				combobox._optionListHeight = null;
				combobox._optionListWidth = null;
				combobox.initializeOptionList();
			}
		});
	});
</script>
```
As you can see, I use the existing `initializeXXX()` methods to re-initialize ComboBox parts.

**Comment by PeterWoods16** [comment]
Hi

The optionList issue was part of the original post.

_Specifically, the corrupted combo2 has zero sized height and width for the button **and list** aspects of the control._  

What I have now is...

[Html4.txt](https://github.com/DevExpress/AjaxControlToolkit/files/666062/Html4.txt)

I plan to experiment with the Sys.Application.add_load part to see if I can get this out of my HTML and into my standard externally referenced JS file.

When installed the code crashed on the final statement...

![image](https://cloud.githubusercontent.com/assets/24647830/21386088/03c1f0b8-c769-11e6-97e3-46e57d9ae03c.png)

So to fix that I found the following...

http://stackoverflow.com/questions/2333297/microsoft-jscript-runtime-error-sys-invalidoperationexception-two-components-w#2753288

Note the final entry in this log.
With my ScriptManager now forced to Release from the default Auto - the above message goes away.

<asp:ScriptManager ID="ScriptManager1" runat="server" ScriptMode="Release"></asp:ScriptManager>

However, I am rather concerned about this.  Looks like I am just suppressing an error?

**Comment by MikhailTymchukDX** [comment]
It's hard to tell what causes this error based only on this exception message.
Would you please provide a sample project that generates this error so we can research it and find the cause?

**Comment by PeterWoods16** [comment]
Here's the project.

[Scratch2.zip](https://github.com/DevExpress/AjaxControlToolkit/files/670630/Scratch2.zip)

In my dev environment you just need to amend the ScriptMode to "Auto" (or remove it) and you get the crash.

<asp:ScriptManager ID="ScriptManager1" runat="server" **ScriptMode="Release"**></asp:ScriptManager>

**Comment by MikhailTymchukDX** [comment]
@PeterWoods16 
Thank you for providing the sample project.

Initialization of an option list is a complex sequence of actions that involves creating a popup behavior component with a certain ID.
When we call it for the second time, the ComboBox tries to create an element with the same ID, which causes this error. 

To fix it, please add these lines before calling `initializeOptionList()`:
```js
if(combobox._popupBehavior) {
	combobox._popupBehavior.remove_showing(combobox._popupShowingHandler);
	combobox._popupBehavior.remove_shown(combobox._popupShownHandler);
	combobox._popupBehavior.remove_hiding(combobox._popupHidingHandler);
	combobox._popupBehavior.dispose();
	combobox._popupBehavior = null;
}
```

**Comment by PeterWoods16** [comment]
Thanks for all your help on this.

The final code I shall deploy is as follows (for anyone else with this issue)...

Tested on IE11 and Chrome 55.0.2883.87 m (64 bit)

[js.txt](https://github.com/DevExpress/AjaxControlToolkit/files/671696/js.txt)

**Comment by PeterWoods16** [comment]
Cheers
